### PR TITLE
OMPI/COLL: Add allgather, reduce coll ucc

### PR DESCRIPTION
**Pull request by lappazos**
Add support for Allgather & Allgatherv. Change Alltoall according to UCC counts API change.

### CODE REVIEW

**Comment by bwbarrett** [comment]
ok to test

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ab0635a8ef1f4f31ed8448747ce307d5

**Comment by jjhursey** [comment]
bot:ibm:retest

File: ompi/mca/coll/ucc/coll_ucc.h [hunk 1]
```diff
@@ -87,6 +88,14 @@ struct mca_coll_ucc_module_t {
     mca_coll_base_module_t*              previous_alltoallv_module;
     mca_coll_base_module_ialltoallv_fn_t previous_ialltoallv;
     mca_coll_base_module_t*              previous_ialltoallv_module;
+    mca_coll_base_module_alltoall_fn_t   previous_allgather;
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc.h:91]: alltoall_fn_t -> allgather_fn_t
ialltoall_fn_t ->iallgather_fn_t

File: ompi/mca/coll/ucc/coll_ucc_allgather.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    COLL_UCC_GET_REQ(coll_req);
+    COLL_UCC_CHECK(mca_coll_ucc_allgather_init(sbuf, scount, sdtype,
+                                              rbuf, rcount, rdtype,
+                                              ucc_module, &req, coll_req));
+    COLL_UCC_CHECK(ucc_collective_post(req));
+    *request = &coll_req->super;
+    return OMPI_SUCCESS;
+fallback:
+    UCC_VERBOSE(3, "running fallback iallgather");
+    return ucc_module->previous_iallgather(sbuf, scount, sdtype, rbuf, rcount, rdtype,
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_allgather.c:104]: we have a leak of coll_req here (it was in the original coll/ucc implementation). If we go to fallback then coll_req might leak. Please set coll_req to NULL in line 92. Then before line 104 check:
if (coll_req) {
mca_coll_ucc_req_free(coll_req);
}
"previous_iallgather" will allocate the request itself.

Also please check & apply the same fix to ALL other non-blocking coll/ucc collectives

File: ompi/mca/coll/ucc/coll_ucc_allgatherv.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        UCC_VERBOSE(5, "ompi_datatype is not supported: dtype = %s",
+                    (COLL_UCC_DT_UNSUPPORTED == ucc_sdt) ?
+                    sdtype->super.name : rdtype->super.name);
+        goto fallback;
+    }
+
+    ucc_coll_args_t coll = {
+        .mask      = UCC_COLL_ARGS_FIELD_FLAGS,
+        .coll_type = UCC_COLL_TYPE_ALLGATHERV,
+        .flags = UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER |
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_allgatherv.c:35]: how can you assume it is that the buffers are CONTIG? i think it is not correct unless you perform O(N) check.

**Review by vspetrov — CHANGES_REQUESTED** [review]

**Comment by gpaulsen** [comment]
@lappazos Could you please sign off the commits in this Pull Request using the `-s` flag to `git commit`?  By doing so, you and your company agree to https://github.com/open-mpi/ompi/wiki/Administrative-rules#contributors-declaration.  This will also allow the GitHub Action CI / Git commit checker to pass.

File: ompi/mca/coll/ucc/coll_ucc_reduce.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+ * Additional copyrights may follow
+ *
+ */
+
+#include "coll_ucc_common.h"
+
+static inline ucc_status_t mca_coll_ucc_reduce_init(const void *sbuf, void *rbuf, int count,
+                                                    struct ompi_datatype_t *dtype,
+                                                    struct ompi_op_t *op, int root,
+													mca_coll_ucc_module_t *ucc_module,
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce.c:15]: alignment

**Review by vspetrov — APPROVED** [review]

**Review by janjust — APPROVED** [review]

**Comment by lappazos** [comment]
@gpaulsen any idea why the check is stuck?

**Comment by awlauria** [comment]
@lappazos fixed.

**Comment by janjust** [comment]
@lappazos can you please open up a v5.0 of this PR

**Comment by lappazos** [comment]
> @lappazos can you please open up a v5.0 of this PR (please remember to use the -s -x arguments to add the cherry-pick and signoff info)

@janjust @awlauria should I cherry-pick the commit from my own branch (the commit i pushed here), or the new merge commit from master?

**Comment by lappazos** [comment]
> @lappazos can you please open up a v5.0 of this PR (please remember to use the -s -x arguments to add the cherry-pick and signoff info)

Done

**Pull request merged**